synapseclient/api/user_services.py

```diff
@@ -226,3 +230,107 @@ async def _get_certified_passing_record(
         uri=f"/user/{userid}/certifiedUserPassingRecord"
     )
     return response
+
+
+async def get_user_by_principal_id_or_name(
+    principal_id: Optional[Union[str, int]] = None,
+    *,
+    synapse_client: Optional["Synapse"] = None,
+) -> int:
+    """
+    Given either a string, int or None finds the corresponding user where None implies PUBLIC.
+
+    Arguments:
+        principal_id: Identifier of a user or group. '273948' is for all registered Synapse users
+                     and '273949' is for public access. None implies public access.
+        synapse_client: If not passed in and caching was not disabled by
+                `Synapse.allow_client_caching(False)` this will use the last created
+                instance from the Synapse class constructor.
+
+    Returns:
+        The integer ID of the user.
+
+    Raises:
+        SynapseError: If the user cannot be found or is ambiguous.
+
+    Example: Get user ID by principal ID
+        Get the user ID for a given principal ID.
+
+        ```python
+        import asyncio
+        from synapseclient import Synapse
+        from synapseclient.api import get_user_by_principal_id_or_name
+
+        syn = Synapse()
+        syn.login()
+
+        async def main():
+            # Get public user ID
+            public_id = await get_user_by_principal_id_or_name(principal_id=None)
+            print(f"Public user ID: {public_id}")
+
+            # Get user ID by username
+            user_id = await get_user_by_principal_id_or_name(principal_id="username")
+            print(f"User ID for 'username': {user_id}")
+
+        asyncio.run(main())
+        ```
+    """
+    from synapseclient import PUBLIC, Synapse
+
+    client = Synapse.get_client(synapse_client=synapse_client)
+
+    if principal_id is None or principal_id == "PUBLIC":
+        return PUBLIC
+    try:
+        return int(principal_id)
```

Comment:
I'm not sure how often this occurs, but I think usernames can also be integers, which can complicate things

Reply:
That is right that they can be ints. This logic in the API classes was mostly copied from the `client.py` script, for this method:

https://github.com/Sage-Bionetworks/synapsePythonClient/blob/babe7857aceef55df267eeb684b6ba801202f862/synapseclient/client.py#L3669-L3707

I decided just to keep these quirks as is.

Where the user is meant to start the chain that leads them to this function clearly indicates that they should be providing the Principal ID, and not the Username:
https://github.com/Sage-Bionetworks/synapsePythonClient/blob/babe7857aceef55df267eeb684b6ba801202f862/synapseclient/models/mixins/access_control.py#L300-L309
https://github.com/Sage-Bionetworks/synapsePythonClient/blob/babe7857aceef55df267eeb684b6ba801202f862/synapseclient/models/mixins/access_control.py#L256-L262